### PR TITLE
[DOC] Enhanced RDoc for Regexp

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Treats:
- #source
- #inspect
- #to_s
- #casefold?
- #options
- #names
- #named_captures

Also corrects casing for headings in regexp.rdoc, some of which are/will be linked from re.c; searched for existing links to the modified headings but found none.

### CODE REVIEW

File: re.c [hunk 1]
```diff
@@ -503,14 +503,16 @@ rb_reg_source(VALUE re)
...
+ *    inspect -> string
+ *
+ *  Returns a nicely-formatted string representation of +self+:
+ *
+ *    /ab+c/ix.inspect # => "/ab+c/ix"
  *
- * Produce a nicely formatted string-version of _rxp_. Perhaps surprisingly,
- * <code>#inspect</code> actually produces the more natural version of
- * the string than <code>#to_s</code>.
+ *  Contrast with Regexp#to_s:
```
Comment (jeremyevans) [inline re.c:513]: I realize it was already there, but I would drop this part and the example below, as `Regexp#to_s` behavior is covered in it's own documentation.

File: re.c [hunk 2]
```diff
@@ -527,22 +529,31 @@ static VALUE rb_reg_str_with_term(VALUE re, int term);
...
+ *    r2 = /#{s0}/        # => /(?ix-m:ab+c)/
+ *
+ *  Note that +r1+ and +r2+ are not equal to +r0+
+ *  because their original strings are different:
+ *
+ *    r0 == r1  # => false
+ *    r0.source # => "ab+c"
+ *    r1.source # => "(?ix-m:ab+c)"
+ *
+ *  Contrast with Regexp#inspect:
```
Comment (jeremyevans) [inline re.c:553]: Likewise, I don't see a need for this paragraph and the example below.

File: re.c [hunk 3]
```diff
@@ -732,25 +745,39 @@ rb_reg_casefold_p(VALUE re)
...
+ *    /foo/m.options   # => 4
+ *    /foo/mix.options # => 7
+ *
+ *  Note that additional bits may be set in the returned integer;
+ *  these are maintained internally internally in +self+,
+ *  are ignored if passed to Regexp.new, and may be ignored by the caller:
  *
  *  Returns the set of bits corresponding to the options used when
- *  creating this Regexp (see Regexp::new for details. Note that
+ *  creating this regexp (see Regexp::new for details. Note that
```
Comment (jeremyevans) [inline re.c:771]: `s/details/details)/`

**Review by jeremyevans — CHANGES_REQUESTED** [review]
One very minor punctuation fix requested.  A couple other inline suggestions, but those are up to you.  After the punctuation fix, can be committed without further review.

File: re.c (hunk 1 above)
Comment (BurdetteLamar) [reply re.c:513]: @jeremyevans, change to Related:?

Comment (jeremyevans) [reply re.c:513]: Sure, `Related: #to_s` seems fine.

**Pull request merged**